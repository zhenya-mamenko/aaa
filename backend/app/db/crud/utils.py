import json
import re

from pydantic import BaseModel as _BaseModel

from app.db.sqlite import (
    execute_query as _execute_query,
    get_database_conn as _get_database_conn,
    read_query as _read_query,
)


def create_parameters_from_model(
    model: _BaseModel,
    /,
    exclude_fields: tuple[str] | None = None,
    only_fields: tuple[str] | None = None,
    get_max_data: tuple | None = None,
) -> tuple:
    """
    Create SQL parameters from a Pydantic model.

    Args:
        model (_BaseModel): The Pydantic model instance.
        exclude_fields (tuple[str] | None, optional): Fields to exclude from the model. Defaults to None.
        only_fields (tuple[str] | None, optional): Fields to include from the model. Defaults to None.
        get_max_data (tuple | None, optional): Tuple containing table name and field to get max value. Defaults to None.

    Returns:
        tuple: A tuple containing the SQL template, fields, parameters, and optionally a max query.
    """

    exclude_fields = exclude_fields or ()
    model_dump = {
        k: json.dumps(v) if type(v).__name__ in ["list", "dict"] else v for k, v in model.model_dump().items()
    }

    if only_fields:
        fields = [field for field in only_fields if field in model_dump]
        if len(fields) != len(only_fields):
            raise ValueError(f"Fields={set(fields) ^ set(only_fields)} are not in the model.")
    else:
        fields = [field for field, value in model_dump.items() if field not in exclude_fields and value is not None]
    parameters = tuple([model_dump[field] for field in fields])
    template = ", ".join(["?" for _ in parameters])

    if get_max_data:
        table_name, field = get_max_data
        if field in fields:
            raise ValueError(f"Field {field} must be excluded from the model.")
        fields.append(field)
        template += f", (SELECT COALESCE(MAX({field}), 0) + 1 FROM {table_name})"
        max_query = f"SELECT MAX({field}) as {field} FROM {table_name};"
        return template, fields, parameters, max_query

    return template, fields, parameters


def insert_data_from_model(
    *, table_name: str, id_field: tuple | str, model: _BaseModel, auto_increment_id: bool = True
) -> int | str | None:
    """
    Insert data into a table from a Pydantic model.

    Args:
        table_name (str): The name of the table to insert data into.
        id_field (tuple | str): The field(s) representing the ID.
        model (_BaseModel): The Pydantic model instance.
        auto_increment_id (bool, optional): Whether the ID is auto-incremented. Defaults to True.

    Returns:
        int | str | None: The ID of the inserted row if auto_increment_id is True, otherwise None.
    """

    if auto_increment_id:
        if isinstance(id_field, tuple):
            raise ValueError("auto_increment_id must be False if id_field is a tuple.")
        template, fields, parameters, q = create_parameters_from_model(model, get_max_data=(table_name, id_field))
    else:
        template, fields, parameters = create_parameters_from_model(model)

    with _get_database_conn() as conn:
        query = f"""
            INSERT INTO {table_name} (
                {", ".join(fields)}
            ) VALUES (
                {template}
            );
        """
        _execute_query(query, parameters=parameters, conn=conn)
        if auto_increment_id:
            id = _read_query(q, conn=conn)[0][id_field]
            return id

    return None


def update_data_from_model(*, table_name: str, id_field: tuple | str, model: _BaseModel) -> bool:
    """
    Update data in a table from a Pydantic model.

    Args:
        table_name (str): The name of the table to update data in.
        id_field (tuple | str): The field(s) representing the ID.
        model (_BaseModel): The Pydantic model instance.

    Returns:
        bool: True if the update was successful, False otherwise.
    """

    if isinstance(id_field, str):
        id_field = (id_field,)

    _, fields, parameters = create_parameters_from_model(model, exclude_fields=id_field)

    query = f"""
        UPDATE {table_name}
        SET
            {", ".join([f"{field} = ?" for field in fields])}
        WHERE
            {' AND '.join([f'{f} = ?' for f in id_field])};
    """
    parameters += tuple([getattr(model, f) for f in id_field])
    result = _execute_query(query, parameters=parameters)
    return result == 1


def delete_data_by_id(*, table_name: str, id_field: tuple | str, id_value: int | str | tuple) -> bool:
    """
    Delete data from a table by ID.

    Args:
        table_name (str): The name of the table to delete data from.
        id_field (tuple | str): The field(s) representing the ID.
        id_value (int | str | tuple): The value(s) of the ID to delete.

    Returns:
        bool: True if the deletion was successful, False otherwise.
    """

    if not isinstance(id_field, tuple):
        id_field=(id_field, )
    query = f"""
        DELETE FROM {table_name}
        WHERE
            {' AND '.join([f'{f} = ?' for f in id_field])};
    """
    if not isinstance(id_value, tuple):
        id_value=(id_value, )
    result = _execute_query(query, parameters=id_value)

    return result == 1


def extract_create_stmts_objects(sql_file_path: str, object_type: str) -> list[str]:
    """
    Extract the CREATE statements objects from a SQL file.

    Args:
        sql_file_path (str): The path to the SQL file.
        object_type (str): The type of objects to extract from the SQL file.

    Returns:
        list: A list of dictionaries containing the objects.
    """

    try:
        with open(sql_file_path, "r") as f:
            sql = f.read()
    except FileNotFoundError:
        raise FileNotFoundError(f"File not found: {sql_file_path}")

    re_template = fr"CREATE\s+{object_type}\s+(IF\s+NOT\s+EXISTS\s+)*(\S+)"
    pattern = re.compile(re_template)
    result = []
    for m in pattern.finditer(sql):
        result.append(m.group(2))

    return result
