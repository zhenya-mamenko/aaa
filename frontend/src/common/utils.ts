import { type AppElement } from "@/types/common";
import { elements } from "@/common/const";


export const isPlainObject: Function = (o: any) => Boolean(
  o && o.constructor && o.constructor.prototype && o.constructor.prototype.hasOwnProperty("isPrototypeOf")
);


export const flattenObject: Function = (obj: any, keys = []) => {
  return Object.keys(obj).reduce((acc: any, key: any) => {
    return Object.assign(acc, isPlainObject(obj[key])
      ? flattenObject(obj[key], keys.concat(key))
      : { [keys.concat(key).join(".")]: obj[key] }
    )
  }, {})
}


export const loadParams: Function = (key: string) => {
  const savedParams = localStorage.getItem(key)
  if (savedParams) {
    return JSON.parse(savedParams)
  }
  return null
}


export const saveParams: Function = (key: string, params: any) => {
  localStorage.setItem(key, JSON.stringify(params))
}


export const deleteParams: Function = (key: string) => {
  localStorage.removeItem(key)
}


export const removeReactivity: Function = (o: any) => {
  return JSON.parse(JSON.stringify(o))
}

export const findElement: Function = (key: string): AppElement | undefined => {
  return elements.find((element: AppElement) => element.key === key);
}


// String.formatUnicorn polyfill from https://stackoverflow.com/a/18234317
export const format: Function = function(str: string, ...rest: any[]): string {
    if (rest.length) {
      var t = typeof rest[0];
      var key;
      var args = ("string" === t || "number" === t) ? Array.prototype.slice.call(rest) : rest[0];
      for (key in args) {
        str = str.replace(new RegExp("\\{" + key + "\\}", "gi"), args[key]);
      }
    }
    return str;
  };


export const parseAxiosError: Function = (error: any): string | any[] => {
  if (error.response?.data?.detail !== undefined) {
    let message = "";
    const detail = error.response.data.detail;
    if (typeof detail === "string" && detail.length > 0) {
      message = `messages.backend.${detail.split(":")[0].replace(/ /g, "_")}`;
    }
    return [
      "messages.backend.request_error",
      error.response.status,
      message,
    ];
  } else if (!error.response) {
    return "messages.backend.no_reponse";
  } else {
    return error.message || "messages.backend.unknown_error";
  }
}


export const calculateReplenishment: Function = (
  list: any[],
  summa: number,
  percentileKey: string = "structure_percentile",
  resultKey: string = "calculated"
): void => {
  list.forEach(item => {
    (item as any)[resultKey] = Math.round(summa * item[percentileKey] / 1000);
  });
}


export const calculateRebalancing: Function = (
  list: any[],
  summa: number,
  amountKey: string = "amount",
  percentileKey: string = "structure_percentile",
  resultKey: string = "calculated"
): void => {
  const total = list.reduce((acc, item) => acc + item[amountKey], 0) + summa * 100;
  list.forEach(item => {
    (item as any)[resultKey] = Math.round(((total * item[percentileKey] / 1000) - item[amountKey]) / 100);
  });
}
