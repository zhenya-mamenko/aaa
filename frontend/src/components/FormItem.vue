<template>
  <v-switch v-if="isBoolean"
    color="primary"
    density="comfortable"
    :data-testid="`switch-${params.title.replace(/\./g, '-').toLowerCase()}`"
    :hide-details="true"
    inset
    :label="t(params.title)"
    v-model="params.value"
  ></v-switch>
  <v-text-field v-else-if="isString || isNumber"
    :counter="isString && params.counter ? params.counter : false"
    :data-testid="`text-field-${params.title.replace(/\./g, '-').toLowerCase()}`"
    hide-spin-buttons
    :label="t(params.title)"
    persistent-hint
    :rules="rules"
    :type="inputType"
    v-model="params.value"
  ></v-text-field>
  <v-select v-else-if="isArray && !isAutoComplete"
    :data-testid="`select-${params.title.replace(/\./g, '-').toLowerCase()}`"
    :items="params.value.map((item: any) => {
      item.title = t(item.title)
      return item;
    })"
    :label="t(params.title)"
    v-model="selectedValue"
  ></v-select>
  <v-autocomplete v-else-if="isArray && isAutoComplete"
    auto-select-first
    :data-testid="`autocomplete-${params.title.replace(/\./g, '-').toLowerCase()}`"
    :items="params.value.map((item: any) => {
      item.title = t(item.title)
      return item;
    })"
    :label="t(params.title)"
    v-model="selectedValue"
  ></v-autocomplete>
</template>

<script lang="ts">
import { defineComponent } from 'vue';
import { t } from "@/common/locale";
import { type Rule } from "@/types/common";


export const rules: Rule = {
  required: () => (v: any) => !!v || t("errors.required"),
  max_length: (max: number) => (v: any) => String(v).length <= max || t("errors.max_length", max),
  number: () => (v: any) => !isNaN(parseFloat(v)) || t("errors.number"),
  min_value: (min: number) => (v: any) => v >= min || t("errors.min_value", min),
  value_range: (min: number, max: number) => (v: any) => v >= min && v <= max || t("errors.value_range", min, max),
}

export default defineComponent({
  name: 'FormItem',
  computed: {
    isArray(): boolean {
      return Array.isArray(this.params.value);
    },
    isBoolean(): boolean {
      return typeof this.params.value === "boolean";
    },
    isDate(): boolean {
      return this.params.type === "date";
    },
    isNumber(): boolean {
      return this.params.type === "number" || typeof this.params.value === "number";
    },
    isString(): boolean {
      return typeof this.params.value === "string";
    },
    isAutoComplete(): boolean {
      return this.params.type === "autocomplete";
    },
    inputType(): string {
      if (this.isDate) return "date";
      return typeof this.params.value === "number" ? "number" : "text";
    },
    rules() {
      if (!this.params.rules) return [];
      const result = [];
      for (const rule of this.params.rules) {
        if (!Array.isArray(rule) && rules[rule]) {
          result.push(rules[rule]());
          continue;
        } else if (!rules[rule[0]]) {
          continue;
        }
        result.push(rules[rule[0]](...rule.slice(1)));
      }
      return result;
    },
    selectedValue: {
      get(): string {
        return this.params.value.find((item: any) => item.selected)?.value;
      },
      set(newValue: string) {
        const selectedItem = this.params.value.find((item: any) => item.value === newValue);
        if (selectedItem) {
          this.params.value.forEach((item: any) => item.selected = false);
          selectedItem.selected = true;
        }
      }
    },
    t() {
      return t;
    },
  },
  props: {
    params: {
      type: Object,
      required: true,
    },
  },
});
</script>
