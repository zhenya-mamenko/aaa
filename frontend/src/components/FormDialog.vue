<template>
  <v-dialog
    max-width="700px"
    :persistent="true"
    v-model="show"
  >
    <v-card
      class="pa-3"
    >
      <v-card-title
        class="ms-2"
        v-html="title"
      />
      <v-card-text>
        <v-form
          class="ma-0 pa-0"
          v-model="isValid"
        >
          <v-container class="ma-0 pa-0">
            <v-row
              v-for="item in results"
              :key="Array.isArray(item) ? (item as Array<any>)[0].title : item.title"
            >
              <v-col
                class="pt-1 pb-1 a"
                cols="12"
                :md="item.cols || (Array.isArray(item) && (item as Array<any>).length > 1 ? ((item as Array<any>)[0].cols || 6) : 12)"
              >
                <form-item
                  :params="Array.isArray(item) ? (item as Array<any>)[0] : item"
                />
              </v-col>
              <v-col v-if="Array.isArray(item) && (item as Array<any>).length > 1"
                class="pt-1 pb-1 b"
                cols="12"
                :md="(item as Array<any>)[1].cols || 6"
              >
                <form-item
                  :params="(item as Array<any>)[1]"
                />
              </v-col>
            </v-row>
            <v-row v-if="columns !== null && Object.keys(results).length > 0">
              <v-col>
                <v-divider class="mt-1 mb-1" />
              </v-col>
            </v-row>
            <v-row v-if="columns !== null">
              <v-col
                class="d-none d-md-inline text-h6"
                cols="3"
              >
                {{ t("settings.columns") }}
              </v-col>
              <v-col>
                <draggable
                  item-key="key"
                  v-model="columns"
                >
                  <template #item="{element}">
                    <div
                      class="border pa-1 ps-3 mb-1 rounded-lg bg-primary-lighten-5"
                      :key="element.title"
                    >
                      <v-switch
                        color="primary"
                        :data-testid="`switch-${element.title}`"
                        density="comfortable"
                        :disabled="element.disabled"
                        :hide-details="true"
                        inset
                        :label="t(element.title)"
                        v-model="element.visible"
                      ></v-switch>
                    </div>
                  </template>
                </draggable>
              </v-col>
            </v-row>
          </v-container>
        </v-form>
      </v-card-text>
      <v-card-actions
        class="pe-4 me-2"
      >
        <v-btn v-if="showResetButton"
          data-testid="form-dialog-reset-button"
          @click="reset"
        >
          <v-icon class="d-inline-block d-md-none" size="x-large">mdi-cog-refresh</v-icon>
          <span class="d-none d-md-inline-block">{{ t("buttons.reset") }}</span>
        </v-btn>
        <v-spacer />
        <v-btn
          color="primary"
          data-testid="form-dialog-save-button"
          :disabled="!isValid"
          @click="save"
        >
          {{ t("buttons.save") }}
        </v-btn>
        <v-btn
          @click="closeDialog"
        >
          {{ t("buttons.close") }}
        </v-btn>
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script lang="ts">
import { defineComponent } from 'vue';
import draggable from "vuedraggable";
import { t } from "@/common/locale";
import { type SettingsColumn, type Settings } from "@/types/common";
import FormItem from "./FormItem.vue";
import { removeReactivity } from '@/common/utils';


export default defineComponent({
  name: "FormDialog",
  components: {
    draggable,
    FormItem,
  },
  data() {
    return {
      isValid: true as boolean,
      columns: [] as SettingsColumn[],
      results: {} as Settings,
      show: false as boolean,
    };
  },
  computed: {
    t() {
      return t;
    },
  },
  emits: [
    "close",
    "reset",
    "save",
  ],
  methods: {
    closeDialog() {
      this.$emit("close");
    },
    parseParams(newValue: Settings) {
      this.columns = !!newValue.columns ? removeReactivity(newValue.columns) : null;
      this.results = Object.fromEntries(
        Object.entries(removeReactivity(newValue)).filter(
          ([key, value]) => key !== "columns" && ((value as any).value !== undefined || Array.isArray(value))
            && (!(value as any).type || (value as any).type !== "hidden")
        )
      );
    },
    reset() {
      this.$emit("reset");
    },
    save() {
      const fe = (res: any, old: any) => {
        return Object.fromEntries(
          Object.entries(res).map(
            ([key, value]) => {
              const p = old[key] as any;
              if (Array.isArray(value)) {
                value = value.map((item: any, index: number) => {
                  return fe(item, p[index]);
                });
              } else if (typeof value === "object") {
                (value as any).title = p.title;
              } else if (key === "title") {
                value = old[key];
              }
              return [key, value];
            }
          )
        )
      }
      const results = {...this.params, ...fe(this.results, this.params)};
      if (this.columns) {
        results.columns = this.columns;
      }
      this.$emit("save", results);
    },
  },
  mounted() {
    this.parseParams(this.params);
  },
  props: {
    title: {
      type: String,
      required: false,
    },
    opened: {
      type: Boolean,
      required: true,
    },
    params: {
      type: Object as () => Settings,
      required: true,
    },
    showResetButton: {
      type: Boolean,
      required: false,
      default: true,
    },
  },
  watch: {
    opened(newValue: boolean) {
      this.show = newValue;
    },
    params(newValue) {
      this.parseParams(newValue);
    },
  },
});
</script>
