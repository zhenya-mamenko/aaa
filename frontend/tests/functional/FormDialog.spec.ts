import { mountComponent, simpleStubs, cardStubs, baseStubs } from "./utils";
// @ts-ignore
import FormItem from "@/components/FormItem.vue";
import draggable from "vuedraggable";
// @ts-ignore
import FormDialog from "@/components/FormDialog.vue";


jest.mock("@/common/locale", () => ({
  t: (key: string) => key,
}));

describe("FormDialog", () => {

  function mountDialog(props: any, data: any) {
    const p = {
      prop1: { value: true, title: "prop1", },
      prop2: { value: "test", title: "prop2", },
    };
    const params = {
      columns: [
        { "key": "column1", "visible": true, disabled: true, title: "column1", },
        { "key": "column2", "visible": false, disabled: false, title: "column2", },
      ],
      ...p
    };
    return mountComponent(
      FormDialog,
      {
        simpleStubs: [...simpleStubs, ],
        stubs: [...cardStubs, ...baseStubs, ],
      },
      {
        opened: true,
        title: "Test",
        params,
        ...props,
      },
      {
        columns: params.columns,
        results: p,
        ...data,
      },
      {
        components: {
          draggable,
          FormItem,
        }
      }
    );
  }

  it("Should render", () => {
    const wrapper = mountDialog({}, {});

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Should cointain the right elements", () => {
    const wrapper = mountDialog({}, {});

    const input = wrapper.get("[data-testid='text-field-prop2']");
    expect(input.attributes("label")).toEqual("prop2");
    expect(input.attributes("modelvalue")).toEqual("test");

    const sw = wrapper.get("[data-testid='switch-prop1']");
    expect(sw.attributes("label")).toEqual("prop1");
    expect(sw.attributes("modelvalue")).toEqual("true");

    expect(wrapper.text()).toContain("settings.columns");

    const columns = wrapper.findAll("[data-testid^='switch-column']");
    expect(columns).toHaveLength(2);
    expect(columns[0].attributes("disabled")).toEqual("true");
    expect(columns[0].attributes("modelvalue")).toEqual("true");
    expect(columns[0].attributes("label")).toEqual("column1");
    expect(columns[1].attributes("disabled")).toEqual("false");
    expect(columns[1].attributes("modelvalue")).toEqual("false");
    expect(columns[1].attributes("label")).toEqual("column2");
  });

  it("Should show/hide the reset button", () => {
    const wrapper = mountDialog({}, {});

    const button = wrapper.get("[data-testid='form-dialog-reset-button']");
    expect(button.isVisible()).toEqual(true);

    const wrapper2 = mountDialog({ showResetButton: false }, {});
    expect(wrapper2.findAll("[data-testid='form-dialog-reset-button']")).toHaveLength(0);
  });

  it("Should emit the right data on save", () => {
    const wrapper = mountDialog({}, {});

    const button = wrapper.get("[data-testid='form-dialog-save-button']");
    button.trigger("click");

    const save = wrapper.emitted("save");
    expect(save).not.toBeUndefined();
    expect((save as any)[0]).toStrictEqual([{
        prop1: { value: true, title: "prop1", },
        prop2: { value: "test", title: "prop2", },
        columns: [
          { "key": "column1", "visible": true, disabled: true, title: "column1", },
          { "key": "column2", "visible": false, disabled: false, title: "column2", },
        ],
      }]);
  });

});
