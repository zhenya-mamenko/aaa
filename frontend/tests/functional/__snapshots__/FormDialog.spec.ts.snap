// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormDialog Should render 1`] = `
"<div max-width="700px" persistent="true" modelvalue="false">
  <div class="pa-3">
    <div class="ms-2">Test</div>
    <div>
      <form class="ma-0 pa-0" modelvalue="true">
        <div class="v-container v-locale--is-ltr ma-0 pa-0">
          <div class="v-row">
            <div class="v-col-md-12 v-col-12 pt-1 pb-1 a">
              <v-switch-stub color="primary" density="comfortable" data-testid="switch-prop1" hide-details="true" inset="" label="prop1" modelvalue="true"></v-switch-stub>
            </div>
            <!--v-if-->
          </div>
          <div class="v-row">
            <div class="v-col-md-12 v-col-12 pt-1 pb-1 a">
              <v-text-field-stub counter="false" data-testid="text-field-prop2" hide-spin-buttons="" label="prop2" persistent-hint="" rules="" type="text" modelvalue="test"></v-text-field-stub>
            </div>
            <!--v-if-->
          </div>
          <div class="v-row">
            <div class="v-col">
              <v-divider-stub class="mt-1 mb-1"></v-divider-stub>
            </div>
          </div>
          <div class="v-row">
            <div class="v-col v-col-3 d-none d-md-inline text-h6">settings.columns</div>
            <div class="v-col">
              <div>
                <div class="border pa-1 ps-3 mb-1 rounded-lg bg-primary-lighten-5" data-draggable="true">
                  <v-switch-stub color="primary" data-testid="switch-column1" density="comfortable" disabled="true" hide-details="true" inset="" label="column1" modelvalue="true"></v-switch-stub>
                </div>
                <div class="border pa-1 ps-3 mb-1 rounded-lg bg-primary-lighten-5" data-draggable="true">
                  <v-switch-stub color="primary" data-testid="switch-column2" density="comfortable" disabled="false" hide-details="true" inset="" label="column2" modelvalue="false"></v-switch-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
    <div class="pe-4 me-2">
      <v-btn-stub data-testid="form-dialog-reset-button"></v-btn-stub>
      <v-spacer-stub></v-spacer-stub>
      <v-btn-stub color="primary" data-testid="form-dialog-save-button" loading="false" disabled="false"></v-btn-stub>
      <v-btn-stub></v-btn-stub>
    </div>
  </div>
</div>"
`;
